File: backend/app/schemas/interview.py
from __future__ import annotations

from pydantic import BaseModel, Field


class InterviewStartIn(BaseModel):
    scenario: str = Field(default="job", max_length=24)
    job_role: str = Field(min_length=1, max_length=64)
    difficulty: str = Field(default="medium", max_length=16)
    question_count: int = Field(default=4, ge=2, le=8)
    resume_url: str = Field(default="", max_length=1024)
    resume_profile: dict = Field(default_factory=dict)
    assignment_id: str = Field(default="", max_length=36)
    consent: bool = True


class InterviewSessionBriefOut(BaseModel):
    id: str
    scenario: str
    job_role: str
    job_role_label: str = ""
    difficulty: str
    question_count: int
    status: str
    overall_score: float | None = None
    current_turn: int = 0
    created_at: str = ""
    finished_at: str = ""
    user_id: str = ""
    assignment_id: str = ""
    student_name: str = ""
    review_status: str = ""


class InterviewQuestionOut(BaseModel):
    index: int
    kind: str
    kind_label: str = ""
    question: str
    followup_of: str = ""


class InterviewTurnOut(BaseModel):
    id: str
    turn_index: int
    question: str
    question_kind: str
    transcript: str = ""
    audio_url: str = ""
    frame_urls: list[str] = []
    semantic_score: float | None = None
    prosody_score: float | None = None
    visual_score: float | None = None
    fused_score: float | None = None
    prosody_detail: dict = {}
    feedback: str = ""
    followup_strategy: str = "next"
    duration_sec: float = 0.0


class InterviewReportOut(BaseModel):
    id: str
    session_id: str
    dimension_scores: dict[str, float] = {}
    dimension_labels: dict[str, str] = {}
    key_issues: list[str] = []
    suggestions: list[str] = []
    resource_refs: list[dict] = []
    council_views: dict = {}
    teacher_comment: str = ""
    teacher_score: float | None = None
    review_status: str = "pending"
    degraded_modalities: list[str] = []
    summary: str = ""
    created_at: str = ""


class InterviewSessionDetailOut(InterviewSessionBriefOut):
    class_id: str = ""
    resume_url: str = ""
    resume_profile: dict = {}
    questions: list[InterviewQuestionOut] = []
    turns: list[InterviewTurnOut] = []
    report: InterviewReportOut | None = None
    prep_run_id: str = ""
    prep_intel: dict = {}
    dimension_labels: dict[str, str] = {}


class InterviewResumeOut(BaseModel):
    url: str
    profile: dict = {}
    text_preview: str = ""


class InterviewJobRoleOut(BaseModel):
    key: str
    label: str
    scenario: str
    family: str = ""
    description: str = ""


class InterviewTaskOut(BaseModel):
    assignment_id: str
    title: str
    description: str = ""
    due_at: str = ""
    scenario: str = "job"
    job_role: str = ""
    question_count: int = 4
    difficulty: str = "medium"
    stem: str = ""
    my_status: str = "pending"
    my_score: int | None = None


class InterviewTeacherReviewIn(BaseModel):
    comment: str = Field(default="", max_length=2000)
    score: float | None = Field(default=None, ge=0, le=100)
    status: str = Field(default="reviewed", max_length=24)


class InterviewOverviewOut(BaseModel):
    total: int = 0
    completed: int = 0
    pending_review: int = 0
    avg_score: float | None = None
    job_count: int = 0
    academic_count: int = 0


class InterviewPracticeQuestionOut(BaseModel):
    question: str
    kind: str
    kind_label: str = ""
    scenario: str = "job"
    job_role: str = ""
    job_role_label: str = ""


class InterviewPracticeAnswerIn(BaseModel):
    scenario: str = Field(default="job", max_length=24)
    job_role: str = Field(default="", max_length=64)
    kind: str = Field(default="", max_length=32)
    question: str = Field(min_length=1, max_length=2000)
    transcript: str = Field(default="", max_length=4000)


class InterviewPracticeAnswerOut(BaseModel):
    id: str
    score: float | None = None
    feedback: str = ""
    star_hit: dict[str, bool] = {}
    reasons: list[str] = []
    created_at: str = ""


class InterviewPracticeRecordOut(BaseModel):
    id: str
    scenario: str = "job"
    job_role: str = ""
    job_role_label: str = ""
    kind: str = ""
    kind_label: str = ""
    question: str = ""
    transcript: str = ""
    score: float | None = None
    feedback: str = ""
    star_hit: dict[str, bool] = {}
    created_at: str = ""


class InterviewPortraitLatestOut(BaseModel):
    id: str
    scenario: str
    job_role: str
    job_role_label: str = ""
    overall_score: float | None = None
    created_at: str = ""


class InterviewPortraitScenarioOut(BaseModel):
    count: int = 0
    avg_score: float | None = None
    dimension_avg: dict[str, float] = {}
    dimension_latest: dict[str, float] = {}
    dimension_labels: dict[str, str] = {}
    latest_id: str = ""
    latest_job_role: str = ""
    latest_job_role_label: str = ""


class InterviewPortraitRoleOut(BaseModel):
    job_role: str
    job_role_label: str = ""
    scenario: str = "job"
    count: int = 0
    avg_score: float | None = None


class InterviewPortraitTrendOut(BaseModel):
    id: str
    at: str = ""
    overall_score: float | None = None
    scenario: str = "job"
    job_role_label: str = ""


class InterviewPortraitWeakDimOut(BaseModel):
    key: str
    label: str
    avg: float
    scenario: str


class InterviewPortraitOut(BaseModel):
    session_count: int = 0
    avg_score: float | None = None
    latest: InterviewPortraitLatestOut | None = None
    job: InterviewPortraitScenarioOut = Field(default_factory=InterviewPortraitScenarioOut)
    academic: InterviewPortraitScenarioOut = Field(default_factory=InterviewPortraitScenarioOut)
    by_role: list[InterviewPortraitRoleOut] = []
    trend: list[InterviewPortraitTrendOut] = []
    weak_dims: list[InterviewPortraitWeakDimOut] = []
    loop_counts: dict[str, int] = {}
    recent_refs: list[dict] = []


class InterviewResumeCoachIn(BaseModel):
    text: str = Field(default="", max_length=8000)
    profile: dict = Field(default_factory=dict)
    target_role: str = Field(default="", max_length=64)
    jd: str = Field(default="", max_length=4000)


class InterviewResumeOptimizeOut(BaseModel):
    score: int = 0
    issues: list[str] = []
    rewritten_markdown: str = ""
    ats_keywords: list[str] = []
    degraded: bool = False


class InterviewResumeMatchOut(BaseModel):
    score: int = 0
    matched: list[str] = []
    gaps: list[str] = []
    prep_suggestions: list[str] = []
    recommended_portals: list[dict] = []
    degraded: bool = False


class InterviewResumeDocxIn(BaseModel):
    template_id: str = Field(default="editorial", max_length=64)
    fields: dict = Field(default_factory=dict)
    format: str = Field(default="docx", max_length=16)


class InterviewApplicationIn(BaseModel):
    company: str = Field(default="", max_length=128)
    role: str = Field(default="", max_length=128)
    portal_url: str = Field(default="", max_length=1024)
    status: str = Field(default="wishlist", max_length=24)
    notes: str = Field(default="", max_length=2000)


class InterviewApplicationPatch(BaseModel):
    company: str | None = Field(default=None, max_length=128)
    role: str | None = Field(default=None, max_length=128)
    portal_url: str | None = Field(default=None, max_length=1024)
    status: str | None = Field(default=None, max_length=24)
    notes: str | None = Field(default=None, max_length=2000)


class InterviewApplicationOut(BaseModel):
    id: str
    company: str = ""
    role: str = ""
    portal_url: str = ""
    status: str = "wishlist"
    notes: str = ""
    applied_at: str = ""
    created_at: str = ""
    updated_at: str = ""

File: frontend/src/api/interview.ts
import { apiDelete, apiGet, apiPatch, apiPost, apiPostForm } from './client';

export interface InterviewJobRole {
  key: string;
  label: string;
  scenario: string;
  family: string;
  description: string;
}

export interface InterviewSessionBrief {
  id: string;
  scenario: string;
  job_role: string;
  job_role_label: string;
  difficulty: string;
  question_count: number;
  status: string;
  overall_score: number | null;
  current_turn: number;
  created_at: string;
  finished_at: string;
  user_id?: string;
  assignment_id?: string;
  student_name?: string;
  review_status?: string;
}

export interface InterviewQuestion {
  index: number;
  kind: string;
  kind_label: string;
  question: string;
}

export interface InterviewTurn {
  id: string;
  turn_index: number;
  question: string;
  question_kind: string;
  transcript: string;
  audio_url: string;
  frame_urls: string[];
  semantic_score: number | null;
  prosody_score: number | null;
  visual_score: number | null;
  fused_score: number | null;
  prosody_detail: {
    speech_rate?: number;
    filler_count?: number;
    pause_ratio?: number;
    duration_sec?: number;
    char_count?: number;
    reasons?: string[];
  };
  feedback: string;
  followup_strategy: string;
  duration_sec: number;
}

export interface InterviewPrepIntel {
  job?: { summary?: string; skills?: string[]; risks?: string[] };
  profile?: { summary?: string; hooks?: string[] };
  topics?: string[];
}

export interface InterviewReport {
  id: string;
  session_id: string;
  dimension_scores: Record<string, number>;
  dimension_labels: Record<string, string>;
  key_issues: string[];
  suggestions: string[];
  resource_refs: Record<string, unknown>[];
  council_views: Record<string, unknown>;
  teacher_comment: string;
  teacher_score: number | null;
  review_status: string;
  degraded_modalities: string[];
  summary: string;
  created_at: string;
}

export interface InterviewPortraitLatest {
  id: string;
  scenario: string;
  job_role: string;
  job_role_label: string;
  overall_score: number | null;
  created_at: string;
}

export interface InterviewPortraitScenario {
  count: number;
  avg_score: number | null;
  dimension_avg: Record<string, number>;
  dimension_latest: Record<string, number>;
  dimension_labels: Record<string, string>;
  latest_id: string;
  latest_job_role: string;
  latest_job_role_label: string;
}

export interface InterviewPortrait {
  session_count: number;
  avg_score: number | null;
  latest: InterviewPortraitLatest | null;
  job: InterviewPortraitScenario;
  academic: InterviewPortraitScenario;
  by_role: Array<{
    job_role: string;
    job_role_label: string;
    scenario: string;
    count: number;
    avg_score: number | null;
  }>;
  trend: Array<{
    id: string;
    at: string;
    overall_score: number | null;
    scenario: string;
    job_role_label: string;
  }>;
  weak_dims: Array<{
    key: string;
    label: string;
    avg: number;
    scenario: string;
  }>;
  loop_counts: Record<string, number>;
  recent_refs: Record<string, unknown>[];
}

export interface InterviewSessionDetail extends InterviewSessionBrief {
  class_id: string;
  resume_url: string;
  resume_profile: Record<string, unknown>;
  questions: InterviewQuestion[];
  turns: InterviewTurn[];
  report: InterviewReport | null;
  prep_run_id: string;
  prep_intel: InterviewPrepIntel;
  dimension_labels: Record<string, string>;
}

export interface InterviewPracticeQuestion {
  question: string;
  kind: string;
  kind_label: string;
  scenario: string;
  job_role: string;
  job_role_label: string;
}

export interface InterviewPracticeAnswer {
  id: string;
  score: number | null;
  feedback: string;
  star_hit: Record<string, boolean>;
  reasons: string[];
  created_at: string;
}

export interface InterviewPracticeRecord {
  id: string;
  scenario: string;
  job_role: string;
  job_role_label: string;
  kind: string;
  kind_label: string;
  question: string;
  transcript: string;
  score: number | null;
  feedback: string;
  star_hit: Record<string, boolean>;
  created_at: string;
}

export interface InterviewResumeResult {
  url: string;
  profile: Record<string, unknown>;
  text_preview: string;
}

export interface InterviewStreamEvent {
  role: string;
  type: string;
  content: string;
  payload: Record<string, unknown>;
}

export const fetchInterviewRoles = (scenario = '') => {
  const q = scenario ? `?scenario=${encodeURIComponent(scenario)}` : '';
  return apiGet<InterviewJobRole[]>(`/api/interview/job-roles${q}`);
};

export const uploadInterviewResume = (file: File) => {
  const form = new FormData();
  form.append('file', file);
  return apiPostForm<InterviewResumeResult>('/api/interview/resume', form);
};

export const createInterviewSession = (body: {
  scenario: string;
  job_role: string;
  difficulty?: string;
  question_count?: number;
  resume_url?: string;
  resume_profile?: Record<string, unknown>;
  assignment_id?: string;
  consent?: boolean;
}) => apiPost<InterviewSessionBrief>('/api/interview/sessions', body, { timeoutMs: 120_000 });

export interface InterviewTask {
  assignment_id: string;
  title: string;
  description: string;
  due_at: string;
  scenario: string;
  job_role: string;
  question_count: number;
  difficulty: string;
  stem: string;
  my_status: string;
  my_score: number | null;
}

export const fetchInterviewSessions = () => apiGet<InterviewSessionBrief[]>('/api/interview/sessions');

export const fetchInterviewTasks = () => apiGet<InterviewTask[]>('/api/interview/tasks');

export const deleteInterviewSession = (id: string) =>
  apiDelete<{ ok: boolean }>(`/api/interview/sessions/${encodeURIComponent(id)}`);

export const fetchInterviewSession = (id: string) =>
  apiGet<InterviewSessionDetail>(`/api/interview/sessions/${encodeURIComponent(id)}`);

export const fetchInterviewReport = (id: string) =>
  apiGet<InterviewReport>(`/api/interview/reports/${encodeURIComponent(id)}`);

export const fetchInterviewPortrait = () => apiGet<InterviewPortrait>('/api/interview/portrait');

export const fetchPracticeQuestion = (params: { scenario: string; job_role: string; kind?: string }) => {
  const q = new URLSearchParams({
    scenario: params.scenario,
    job_role: params.job_role,
    kind: params.kind || '',
  });
  return apiGet<InterviewPracticeQuestion>(`/api/interview/practice/question?${q.toString()}`);
};

export const submitPracticeAnswer = (body: {
  scenario: string;
  job_role: string;
  kind: string;
  question: string;
  transcript: string;
}) => apiPost<InterviewPracticeAnswer>('/api/interview/practice/answer', body, { timeoutMs: 60_000 });

export const fetchPracticeHistory = () =>
  apiGet<InterviewPracticeRecord[]>('/api/interview/practice/history');

export interface CareerPortal {
  id: string;
  name: string;
  group: string;
  url: string;
  intern_url: string;
  note: string;
  accent?: string;
  logo_host?: string;
}

export interface CareerWindow {
  id: string;
  title: string;
  when: string;
  season: string;
  portal_ids: string[];
  companies: string[];
  note: string;
}

export interface ResumeTemplateMeta {
  id: string;
  name: string;
  description: string;
  suitable: string;
  accent: string;
  allow_photo?: boolean;
  tier?: string;
}

export interface ResumeOpenSourceLink {
  id: string;
  name: string;
  license: string;
  url: string;
  note: string;
}

export interface CareerQuestion {
  id: string;
  company_id: string;
  company: string;
  job_role: string;
  kind: string;
  question: string;
}

export interface InterviewApplication {
  id: string;
  company: string;
  role: string;
  portal_url: string;
  status: string;
  notes: string;
  applied_at: string;
  created_at: string;
  updated_at: string;
}

export interface ResumeOptimizeResult {
  score: number;
  issues: string[];
  rewritten_markdown: string;
  ats_keywords: string[];
  degraded: boolean;
}

export interface ResumeMatchResult {
  score: number;
  matched: string[];
  gaps: string[];
  prep_suggestions: string[];
  recommended_portals: Array<{ id: string; name: string; url: string }>;
  degraded: boolean;
}

export const fetchCareerPortals = () =>
  apiGet<{ portals: CareerPortal[]; windows: CareerWindow[] }>('/api/interview/career/portals');

export const fetchCareerTemplates = () =>
  apiGet<{ templates: ResumeTemplateMeta[]; open_source: ResumeOpenSourceLink[] }>(
    '/api/interview/career/templates',
  );

export const fetchCareerQuestions = (params?: { company?: string; job_role?: string }) => {
  const q = new URLSearchParams();
  if (params?.company) q.set('company', params.company);
  if (params?.job_role) q.set('job_role', params.job_role);
  const suffix = q.toString() ? `?${q}` : '';
  return apiGet<{ companies: Array<{ id: string; name: string }>; questions: CareerQuestion[] }>(
    `/api/interview/career/questions${suffix}`,
  );
};

export const optimizeInterviewResume = (body: {
  text?: string;
  profile?: Record<string, unknown>;
  target_role?: string;
  jd?: string;
}) => apiPost<ResumeOptimizeResult>('/api/interview/resume/optimize', body, { timeoutMs: 60_000 });

export const matchInterviewResume = (body: {
  text?: string;
  profile?: Record<string, unknown>;
  target_role?: string;
  jd?: string;
}) => apiPost<ResumeMatchResult>('/api/interview/resume/match', body, { timeoutMs: 60_000 });

export async function downloadResumeExport(body: {
  template_id: string;
  fields: Record<string, unknown>;
  format: 'html' | 'docx' | 'md';
}): Promise<void> {
  const token = localStorage.getItem('sparkorbit_token');
  const res = await fetch('/api/interview/resume/export', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(await res.text());
  const blob = await res.blob();
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  const ext = body.format === 'docx' ? 'docx' : body.format;
  a.download = `resume-${String(body.fields.name || 'sparkorbit')}.${ext}`;
  a.click();
  URL.revokeObjectURL(url);
}

export async function fetchResumeHtml(body: {
  template_id: string;
  fields: Record<string, unknown>;
}): Promise<string> {
  const token = localStorage.getItem('sparkorbit_token');
  const res = await fetch('/api/interview/resume/export', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify({ ...body, format: 'html' }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.text();
}

export const fetchInterviewApplications = () =>
  apiGet<InterviewApplication[]>('/api/interview/applications');

export const createInterviewApplication = (body: {
  company: string;
  role?: string;
  portal_url?: string;
  status?: string;
  notes?: string;
}) => apiPost<InterviewApplication>('/api/interview/applications', body);

export const patchInterviewApplication = (
  id: string,
  body: Partial<Pick<InterviewApplication, 'company' | 'role' | 'portal_url' | 'status' | 'notes'>>,
) => apiPatch<InterviewApplication>(`/api/interview/applications/${encodeURIComponent(id)}`, body);

export const deleteInterviewApplication = (id: string) =>
  apiDelete<{ ok: boolean }>(`/api/interview/applications/${encodeURIComponent(id)}`);

export function interviewWsUrl(sessionId: string) {
  const proto = window.location.protocol === 'https:' ? 'wss' : 'ws';
  return `${proto}://${window.location.host}/api/ws/interview/${encodeURIComponent(sessionId)}`;
}

export async function consumeInterviewPrepStream(
  sessionId: string,
  onEvent: (event: InterviewStreamEvent) => void,
): Promise<void> {
  const token = localStorage.getItem('sparkorbit_token');
  const res = await fetch(`/api/interview/sessions/${encodeURIComponent(sessionId)}/prep/stream`, {
    headers: token ? { Authorization: `Bearer ${token}` } : {},
  });
  if (!res.ok || !res.body) throw new Error('面试准备流连接失败');
  const reader = res.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    const parts = buffer.split('\n\n');
    buffer = parts.pop() || '';
    for (const part of parts) {
      const line = part.trim();
      if (!line.startsWith('data:')) continue;
      try {
        onEvent(JSON.parse(line.slice(5).trim()) as InterviewStreamEvent);
      } catch {
        /* ignore */
      }
    }
  }
}
